resources/assets/lib/components/report-form.tsx

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import { Modal } from 'components/modal';
+import { SelectOptions } from 'components/select-options';
+import { isEmpty } from 'lodash';
+import * as _ from 'lodash';
+import * as React from 'react';
+
+const bn = 'report-form';
+const maxLength = 2000;
+
+interface Props {
+  completed: boolean;
+  disabled: boolean;
+  onClose: () => void;
+  onSubmit: ({comments}: {comments: string}) => void;
+  title: string;
+  visible: boolean;
+  visibleOptions?: string[];
+}
+
+interface ReportOption {
+  id: string;
+  text: string;
+}
+
+interface State {
+  comments: string;
+  selectedReason: ReportOption;
+}
+
+export class ReportForm extends React.PureComponent<Props, State> {
+  options = [
+    { id: 'Cheating', text: osu.trans('users.report.options.cheating') },
+    { id: 'MultipleAccounts', text: osu.trans('users.report.options.multiple_accounts') },
+    { id: 'Insults', text: osu.trans('users.report.options.insults') },
+    { id: 'Spam', text: osu.trans('users.report.options.spam') },
+    { id: 'UnwantedContent', text: osu.trans('users.report.options.unwanted_content') },
+    { id: 'Nonsense', text: osu.trans('users.report.options.nonsense') },
+    { id: 'Other', text: osu.trans('users.report.options.other') },
+  ];
+
+  constructor(props: Props) {
+    super(props);
+
+    if (props.visibleOptions) {
+      this.options = _.intersectionWith(this.options, props.visibleOptions, (left, right) => left.id === right);
+    }
+
+    this.state = {
+      comments: '',
+      selectedReason: this.options[0],
+    };
```

Comment:
Preferably we'd want to use mobx and use `Component` instead of `PureComponent` instead of `state` and `setState`.